### DISCUSSION TITLE
Let's use `corepack`!

### DISCUSSION THREAD

**Discussion by ghost** [honojs/hono#2051] [Ideas]
# What is corepack?

> Corepack is a zero-runtime-dependency Node.js script that acts as a bridge between Node.js projects and the package managers they are intended to be used with during development. In practical terms, Corepack lets you use Yarn, npm, and pnpm without having to install them.
>
> from [nodejs/corepack](https://github.com/nodejs/corepack)

# Why?
corepack automatically installs the version of the package manager specified in package.json.
This will prevent problems caused by different versions, and will make version upgrades and package manager changes easier.

# How?
> Corepack is [distributed by default with all recent Node.js versions](https://nodejs.org/api/corepack.html). Run `corepack enable` to install the required Yarn and pnpm binaries on your path.
>
> from [nodejs/corepack](https://github.com/nodejs/corepack)

# What do I want to do?
Use pnpm. (Déjà vu...? #1392)
It's no surprise that cool other repositories are using pnpm now. For example Next.js, Nuxt, Vite, Astro.
([Repositories using pnpm workspace](https://pnpm.io/workspaces#usage-examples))
And it's secure and, most importantly, fast.

> Please wait! pnpm may possibly cause a dependency-related problem. I’ll try to investigate it. Anyway, thanks!
>
> _Originally posted by @yusukebe in https://github.com/honojs/hono/issues/1392#issuecomment-1704639989_

Look at this table.

| Feature | pnpm | Yarn |
| ---     | ---  | ---  |
| Workspace support | ✔️ | ✔️ |
| Isolated `node_modules` | ✔️ - The default | ✔️ |
| Hoisted `node_modules` | ✔️ | ✔️ |
| Autoinstalling peers | ✔️ | ❌ |
| Plug'n'Play | ✔️ | ✔️ - The default |
| Zero-Installs | ❌ | ✔️ |
| Patching dependencies | ✔️ | ✔️ |
| Managing Node.js versions | ✔️ | ❌ |
| Has a lockfile | ✔️ - `pnpm-lock.yaml` | ✔️ - `yarn.lock` |
| Dynamic package execution | ✔️ - Via `pnpm dlx` | ✔️ - Via `yarn dlx` |

(Taken from [Feature Comparison (pnpm)](https://pnpm.io/feature-comparison).)

And the fact that it is used in many major repositories would attest to this table.

---

@yusukebe If you are interested, I'll make a pr right away.

Originally #1392 and #1405 by @yusukebe and @ThatOneBro. Thanks!

**Comment by yusukebe** [comment]
@nabeken5 

Thanks for suggesting. I can't explain this specifically yet, as I commented below, but I have tried and had problems.

https://github.com/honojs/hono/issues/1392#issuecomment-1704639989

I will compare `yarn` and `pnpm` again, so please wait a while.

**Comment by yusukebe** [comment]
As for Corepack, we use Yarn 4 with honojs/middleware etc. It is very good.

**Comment by ghost** [comment]
@yusukebe 
I would like to know what problems you have had. I know there are people in the Hono community who can help you with your problems, and I want to help.
(If you have another reason for not wanting to use pnpm, please say so)

Anyway, I want to introduce corepack with yarn. May I?
